cpg_pipes/types.py

```diff
@@ -0,0 +1,224 @@
+"""
+Wrappers for bioinformatics file types (CRAM, GVCF, FASTQ, etc).
+"""
+
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Union
+
+from . import Path, to_path
+from . import utils
+from hailtop.batch import ResourceGroup, ResourceFile, Batch
+
+logger = logging.getLogger(__file__)
+
+
+class CramPath:
+    """
+    Represents alignment data on a bucket within the pipeline.
+    Includes a path to a CRAM or a BAM file along with a corresponding index,
+    and a corresponding fingerprint path.
+    """
+
+    def __init__(self, path: str | Path, index_path: Path | str | None = None):
+        self.path = to_path(path)
+        self.is_bam = self.path.suffix == '.bam'
+        self.ext = 'cram' if not self.is_bam else 'bam'
+        self.index_ext = 'crai' if not self.is_bam else 'bai'
+        self._index_path = index_path
+        self.somalier_path = to_path(f'{self.path}.somalier')
+
+    def __str__(self) -> str:
+        return str(self.path)
+
+    def __repr__(self) -> str:
+        return f'CRAM({self.path})'
+
+    def exists(self) -> bool:
+        """
+        CRAM file exists.
+        """
+        return self.path.exists()
+
+    @property
+    def index_path(self) -> Path:
+        """
+        Path to the corresponding index
+        """
+        return (
+            to_path(self._index_path)
+            if self._index_path
+            else to_path(f'{self.path}.{self.index_ext}')
+        )
+
+    def resource_group(self, b: Batch) -> ResourceGroup:
+        """
+        Create a Hail Batch resource group
+        """
+        return b.read_input_group(
+            **{
+                self.ext: str(self.path),
+                f'{self.ext}.{self.index_ext}': str(self.index_path),
+            }
+        )
+
+
+class GvcfPath:
+    """
+    Represents GVCF data on a bucket within the pipeline.
+    Includes a path to a GVCF file along with a corresponding TBI index,
+    and a corresponding fingerprint path.
+    """
+
+    def __init__(self, path: Path | str):
+        self.path = to_path(path)
+        self.somalier_path = to_path(f'{self.path}.somalier')
+
+    def __str__(self) -> str:
+        return str(self.path)
+
+    def __repr__(self) -> str:
+        return f'GVCF({self.path})'
+
+    def exists(self) -> bool:
+        """
+        GVCF file exists.
+        """
+        return self.path.exists()
+
+    @property
+    def tbi_path(self) -> Path:
+        """
+        Path to the corresponding index
+        """
+        return to_path(f'{self.path}.tbi')
+
+    def resource_group(self, b: Batch) -> ResourceGroup:
+        """
+        Create a Hail Batch resource group
+        """
+        return b.read_input_group(
+            **{
+                'g.vcf.gz': str(self.path),
+                'g.vcf.gz.tbi': str(self.tbi_path),
+            }
+        )
+
+
+FastqPath = Union[str, Path, ResourceFile]
+
+
+@dataclass
+class FastqPair:
```

Comment:
Do we have a Single FASTQ case - this is relevant for ONT sequencing (RNA seq too?)

Reply:
I haven't explored ONT or RNA yet. Another PR. I really want to avoid the PR branches to live for longer than a day, my brain just can't handle so many diverging codebases.